Plugin/Shineretrofits/ShippingMethod/Model/Carrier/CustomshippingPlugin.php

```diff
@@ -0,0 +1,120 @@
+ * @license http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+namespace Bolt\Boltpay\Plugin\Shineretrofits\ShippingMethod\Model\Carrier;
+
+use Magento\Framework\App\State;
+use Magento\Framework\Webapi\Rest\Request;
+use Magento\Shipping\Model\Rate\ResultFactory;
+use Magento\Quote\Model\Quote\Address\RateResult\MethodFactory;
+use Shineretrofits\ShippingMethod\Model\Carrier\Customshipping as ShineretrofitsCustomshipping;
+
+class CustomshippingPlugin
+{    
+    /**
+     * @var \Magento\Shipping\Model\Rate\ResultFactory
+     */
+    private $_rateResultFactory;
+
+    /**
+     * @var \Magento\Quote\Model\Quote\Address\RateResult\MethodFactory
+     */
+    private $_rateMethodFactory;
+
+    /**
+     * @var \Magento\Framework\App\State
+     */
+    private $_appState;
+
+    /**
+     * @var \Magento\Framework\Webapi\Rest\Request
+     */
+    private $_restRequest;
+
+    public function __construct(
+        State   $appState,
+        Request $restRequest,
+        ResultFactory $rateResultFactory,
+        MethodFactory $rateMethodFactory
+    ) {
+        $this->_appState = $appState;
+        $this->_restRequest = $restRequest;
+        $this->_rateResultFactory = $rateResultFactory;
+        $this->_rateMethodFactory = $rateMethodFactory;
+    }
+    
+    /**
+     * Checks if we do calculation for Bolt plugin.
+     *
+     * @return bool
+     */
+    protected function isBoltCalculation()
+    {
+        if ($this->_appState->getAreaCode() !== 'webapi_rest') {
+            return false;
+        }
+        if (!empty($this->_restRequest)) {
+            $payload = $this->_restRequest->getContent();
+            if (!empty($payload)) {
+                $transaction = json_decode($payload);
+                if (isset($transaction->order->cart->shipments[0]->reference)
+                    && $transaction->order->cart->shipments[0]->reference == 'shineretrofits_shippingmethod_shineretrofits_shippingmethod'
+                ) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * The original method prohibits using Shineretrofits shipping method shipping method for webapi_rest,
+     * so it returns false for Bolt backoffice order.
+     * We need to check if calculation was called for Bolt and return rates if so.
+     */
+    public function afterCollectRates(
+        ShineretrofitsCustomshipping $subject,
+        $result
+    ) {
+        if ($result !== false) {
+            return $result;
+        }        
+        if (!$subject->isActive()) {
+            return false;
+        }
+        if($this->_appState->getAreaCode() == 'frontend'){
+            return false;
+        }
+        if($this->_appState->getAreaCode() == 'webrest_api'){
+            return false;
+        }
+        if (!$this->isBoltCalculation()) {
+            return false;
+        }
```

Comment:
```suggestion
        if ($result !== false) {
	    return $result;
	}        
	if (
	    !$subject->isActive()
	    || $this->_appState->getAreaCode() == \Magento\Framework\App\Area::AREA_FRONTEND
	    || $this->_appState->getAreaCode() == \Magento\Framework\App\Area::AREA_WEBAPI_REST
	    || !$this->isBoltCalculation()
	) {
	    return false;
	}
```

Reply:
Done, but `webrest_api` does not exist in the M2 core, it seems to be a custom AreaCode created by the dev of this merchant.
So I change the code to

```
if (
    !$subject->isActive()
    || $this->_appState->getAreaCode() == \Magento\Framework\App\Area::AREA_FRONTEND
    || $this->_appState->getAreaCode() == 'webrest_api'
    || !$this->isBoltCalculation()
) {
    return false;
}
```

Reply:
@dumega `webapi_rest` exists here
https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/App/Area.php#L23

Are you sure it's not a typo on their end? It's very strange to have a need of crating a custom area code.

Reply:
@danac-gs In their code, it is **webrest_api**, not **webapi_rest** . And I can ask CSM to confirm with the merchant

Reply:
@danac-gs 
Below is the original code

```
if($this->state->getAreaCode() == 'frontend'){
    return false;
}
if($this->state->getAreaCode() == 'webrest_api'){
    return false;
}
if($this->state->getAreaCode() == 'webapi_rest'){
    return false;
}
```

For webapi_rest, I create function isBoltCalculation() to cover the case.
https://github.com/BoltApp/bolt-magento2/pull/1252/files#diff-9440d2cc9ac6b28f4535e7f561ad104c31e427c677cee31d4e248762e4c7622bR64

So it seems that webrest_api is intended. I think we would better follow the same conditions.

Reply:
I agree. Thanks.